source/reduce/reducer.cpp

```diff
@@ -0,0 +1,154 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <sstream>
+
+#include "source/spirv_reducer_options.h"
+
+#include "reducer.h"
+#include "reduction_pass.h"
+
+namespace spvtools {
+namespace reduce {
+
+struct Reducer::Impl {
+  explicit Impl(spv_target_env env)
+      : target_env(env), made_progress_this_round(false) {}
+
+  const spv_target_env target_env;  // Target environment.
+  MessageConsumer consumer;         // Message consumer.
+  InterestingnessFunction interestingness_function;
+  std::vector<std::unique_ptr<ReductionPass>> passes;
+  std::vector<std::unique_ptr<ReductionPass>>::iterator current_pass;
+  bool made_progress_this_round;
+
+  std::vector<uint32_t> ApplyReduction(const std::vector<uint32_t>& binary);
+  bool Finished();
+};
+
+Reducer::Reducer(spv_target_env env) : impl_(MakeUnique<Impl>(env)) {}
+
+Reducer::~Reducer() = default;
+
+void Reducer::SetMessageConsumer(MessageConsumer c) {
+  for (auto& pass : impl_->passes) {
+    pass->SetMessageConsumer(c);
+  }
+  impl_->consumer = std::move(c);
+}
+
+void Reducer::SetInterestingnessFunction(
+        Reducer::InterestingnessFunction interestingness_function) {
+  impl_->interestingness_function = std::move(interestingness_function);
+}
+
+bool Reducer::Run(std::vector<uint32_t>&& binary_in,
+                  std::vector<uint32_t>& binary_out,
+                  spv_const_reducer_options options) const {
+  impl_->current_pass = impl_->passes.begin();
+
+  std::vector<uint32_t> current = std::move(binary_in);
+
+  // Initial state should be interesting.
+  assert(impl_->interestingness_function(current));
+
+  for (uint32_t current_step = 0; current_step < options->step_limit; ++current_step) {
+
+    impl_->consumer(SPV_MSG_INFO, nullptr, {},
+       ("Reduction step " + std::to_string(current_step)).c_str());
+
+    auto reduction_step_result =
+        impl_->ApplyReduction(current);
+
+    if (reduction_step_result.empty()) {
+      impl_->consumer(SPV_MSG_INFO, nullptr, {},
+                      "No more to reduce; stopping.");
+      break;
+    }
+
+    if (impl_->interestingness_function(reduction_step_result)) {
+      // Interesting:
+      impl_->made_progress_this_round = true;
+      impl_->consumer(SPV_MSG_INFO, nullptr, {}, "Reduction step succeeded.");
+      current = std::move(reduction_step_result);
+    } else {
+      impl_->consumer(SPV_MSG_INFO, nullptr, {}, "Reduction step failed.");
+    }
+
+  }
+
+  binary_out = std::move(current);
+
+  return true;
+}
+
+void Reducer::AddReductionPass(
+    std::unique_ptr<ReductionPass>&& reduction_pass) {
+  impl_->passes.push_back(std::move(reduction_pass));
+}
+
+std::vector<uint32_t> Reducer::Impl::ApplyReduction(
+    const std::vector<uint32_t>& binary) {
```

Comment:
I find this function very hard to read.  The state changes that are stored in the class, and take place in `ApplyReduction` and in the caller `Run` mean I have to keep going back and forth between the two to understand what this function is doing.  I have not thought it through fully.  However, I could see something like:

```
bool something_chaged = false;
uint32_t num_changes = 0;
while(something_changed && num_changes < options->step_limit) {
  something_chaged = false;
  for( auto& pass : impl_->passes ) {
    auto maybe_result = (*current_pass)->ApplyReduction(binary);
    if (MadeProgress(maybe_result)) {
      something_changed = true;
      ++num_changes;
      // Plus any other book keeping.
      if (num_changes >= options->step_limit) break;
    }
  }
}
```

This might be easier to follow than have a function that iterates over the passes, but may leave and resume the iteration the next time it is called.

Reply:
In latest commit I have reworked this a lot to move all of the functionality into one method, getting rid of the member variables that were tracking state, and making the various loop exit conditions as explicit as possible.